app/javascript/schools/courses/types/curriculum_editor/TargetChecklistItem.re

```diff
@@ -0,0 +1,196 @@
+
+let actionStringForKind = kind => {
+  switch (kind) {
+  | Files => "Upload Files"
+  | Link => "Attach a Link"
+  | ShortText => "Write Short Text"
+  | LongText => "Write Long Text"
+  | MultiChoice(_choices) => "Choose from a list"
+  };
+};
+
+let kindAsString = kind => {
+  switch (kind) {
+  | Files => "files"
+  | Link => "link"
+  | ShortText => "shortText"
+  | LongText => "longText"
+  | MultiChoice(_choices) => "multiChoice"
+  };
+};
+
+let make = (~title, ~kind, ~optional) => {
+  {title, kind, optional};
+};
+
+let updateTitle = (title, t) => {
+  {...t, title};
+};
+
+let updateKind = (kind, t) => {
+  {...t, kind};
+};
+
+let updateOptional = (optional, t) => {
+  {...t, optional};
+};
+
+let removeItem = (index, list) => {
+  list |> Js.Array.filteri((_item, i) => i != index);
+};
+
+let moveUp = (t, list) => {
+  list |> Array.to_list |> ListUtils.swapUp(t) |> Array.of_list;
+};
+
+let moveDown = (t, list) => {
+  list |> Array.to_list |> ListUtils.swapDown(t) |> Array.of_list;
+};
+
+let copy = (t, list) => {
+  list
+  |> Array.map(item => item == t ? [item, item] : [item])
+  |> ArrayUtils.flatten;
+};
+
+let removeMultichoiceOption = (choiceIndex, t) => {
+  switch (t.kind) {
+  | MultiChoice(choices) =>
+    let updatedChoices =
+      choices
+      |> Array.mapi((i, choice) => i == choiceIndex ? [] : [choice])
+      |> ArrayUtils.flatten;
+    t |> updateKind(MultiChoice(updatedChoices));
+  | Files
+  | Link
+  | ShortText
+  | LongText => t
+  };
+};
+
+let addMultichoiceOption = t => {
+  switch (t.kind) {
+  | MultiChoice(choices) =>
+    let updatedChoices = [|""|] |> Array.append(choices);
+    t |> updateKind(MultiChoice(updatedChoices));
+  | Files
+  | Link
+  | ShortText
+  | LongText => t
+  };
+};
+
+let updateMultichoiceOption = (choiceIndex, newOption, t) => {
+  switch (t.kind) {
+  | MultiChoice(choices) =>
+    let updatedChoices =
+      choices
+      |> Array.mapi((i, choice) => i == choiceIndex ? newOption : choice);
+    t |> updateKind(MultiChoice(updatedChoices));
+  | Files
+  | Link
+  | ShortText
+  | LongText => t
+  };
+};
+
+let createNew = {title: "", kind: LongText, optional: false};
+
+let metaData = kind => {
+  switch (kind) {
+  | MultiChoice(choices) => choices
+  | Files
+  | Link
+  | ShortText
+  | LongText => [||]
+  };
+};
+
+let isFilesKind = t => {
+  switch (t.kind) {
+  | Files => true
+  | MultiChoice(_choices) => false
+  | Link
+  | ShortText
+  | LongText => false
+  };
+};
+
+let isValidChecklistItem = t => {
+  switch (t.kind) {
+  | MultiChoice(choices) =>
+    choices
+    |> Js.Array.filter(choice => choice |> String.trim |> String.length < 1)
+    |> ArrayUtils.isEmpty
+    && t.title
+    |> String.trim
+    |> String.length >= 1
+  | Files
+  | Link
+  | ShortText
+  | LongText => t.title |> String.trim |> String.length >= 1
+  };
+};
+
+let createDefaultChecklist = [|
+  make(~title="Describe your submission", ~kind=LongText, ~optional=false),
+|];
```

Comment:
I wouldn't keep this binding here. This doesn't really have anything to do with a _single_ item, and it's not an operation either.